File: backend/ai/pipeline.go
package ai

import (
	"context"
	"fmt"
	"strings"

	ghclient "github.com/just-pr/backend/github"
)

// PipelineConfig holds configuration for the two-stage pipeline.
type PipelineConfig struct {
	APIKey           string   // Anthropic API key
	SonnetModel      string   // model for specialist + summary calls (defaults to claude-sonnet-4-6)
	FallbackProvider Provider // used when below threshold or triage fails
}

// PipelineProvider implements Provider using two-stage parallel analysis.
type PipelineProvider struct {
	cfg PipelineConfig
}

// NewPipelineProvider creates a PipelineProvider.
// fallback is used for small PRs and on triage failure.
func NewPipelineProvider(cfg PipelineConfig) *PipelineProvider {
	if cfg.SonnetModel == "" {
		cfg.SonnetModel = "claude-sonnet-4-6"
	}
	return &PipelineProvider{cfg: cfg}
}

func (p *PipelineProvider) Name() string {
	return "pipeline/" + p.cfg.SonnetModel
}

// AnalyzePR satisfies the Provider interface. Delegates to fallback.
// Use AnalyzePRFull for the full pipeline with structured PR data.
func (p *PipelineProvider) AnalyzePR(ctx context.Context, userPrompt string, emit func(StreamEvent) error) error {
	return p.cfg.FallbackProvider.AnalyzePR(ctx, userPrompt, emit)
}

// AnalyzePRFull runs the full two-stage pipeline with structured PR data.
func (p *PipelineProvider) AnalyzePRFull(
	ctx context.Context,
	diff string,
	prFiles []ghclient.PRFile,
	fileContents []ghclient.FileContent,
	emit func(StreamEvent) error,
) error {
	if !aboveThreshold(prFiles, diff) {
		return p.cfg.FallbackProvider.AnalyzePR(ctx, UserPrompt(diff, fileContents), emit)
	}

	triage, err := Triage(ctx, p.cfg.APIKey, prFiles)
	if err != nil || len(triage) == 0 {
		return p.cfg.FallbackProvider.AnalyzePR(ctx, UserPrompt(diff, fileContents), emit)
	}

	type workerResult struct {
		events []StreamEvent
		err    error
	}

	numWorkers := len(triage) + 1 // specialists + summary
	resultsCh := make(chan workerResult, numWorkers)

	// Summary call: risk + summary + systems
	go func() {
		events, err := runSummaryCall(ctx, p.cfg.APIKey, p.cfg.SonnetModel, triage, first2000Lines(diff))
		resultsCh <- workerResult{events, err}
	}()

	// Specialist calls: one per category
	for catID, files := range triage {
		go func(cat string, catFiles []string) {
			events, err := RunSpecialist(ctx, p.cfg.APIKey, p.cfg.SonnetModel, cat, catFiles, diff, fileContents)
			resultsCh <- workerResult{events, err}
		}(catID, files)
	}

	// Stream results as workers complete; deduplicate categories by ID.
	var categoryEvents []StreamEvent
	emittedCategories := make(map[string]bool)
	for range numWorkers {
		r := <-resultsCh
		if r.err != nil {
			continue
		}
		for _, ev := range r.events {
			switch ev.Type {
			case "risk", "summary", "systems":
				if err := emit(ev); err != nil {
					return err
				}
			case "category":
				id, _ := ev.Data["id"].(string)
				if id != "" && emittedCategories[id] {
					continue // drop duplicate category
				}
				if id != "" {
					emittedCategories[id] = true
				}
				categoryEvents = append(categoryEvents, ev)
				if err := emit(ev); err != nil {
					return err
				}
			}
		}
	}

	// Recommendation: synthesize from all category summaries
	recEvents, err := runRecommendationCall(ctx, p.cfg.APIKey, p.cfg.SonnetModel, categoryEvents)
	if err == nil {
		for _, ev := range recEvents {
			if err := emit(ev); err != nil {
				return err
			}
		}
	}

	return emit(StreamEvent{Type: "done", Data: map[string]any{}})
}

// aboveThreshold returns true if the PR warrants the pipeline path.
func aboveThreshold(files []ghclient.PRFile, diff string) bool {
	if len(files) >= 10 {
		return true
	}
	return strings.Count(diff, "\n") >= 500
}

// first2000Lines returns at most the first 2000 lines of s.
func first2000Lines(s string) string {
	lines := strings.SplitN(s, "\n", 2002)
	if len(lines) <= 2000 {
		return s
	}
	return strings.Join(lines[:2000], "\n") + "\n[truncated for summary]"
}

// runSummaryCall produces risk, summary, and systems events from a condensed diff.
func runSummaryCall(ctx context.Context, apiKey, model string, triage TriageResult, condensedDiff string) ([]StreamEvent, error) {
	var b strings.Builder
	b.WriteString("# Triage classification\n\n")
	for cat, files := range triage {
		fmt.Fprintf(&b, "- **%s**: %s\n", cat, strings.Join(files, ", "))
	}
	b.WriteString("\n# Diff (first 2000 lines)\n\n```diff\n")
	b.WriteString(condensedDiff)
	b.WriteString("\n```\n")

	return runStreamingCall(ctx, apiKey, model, SummarySystemPrompt(), b.String(), func(ev StreamEvent) bool {
		return ev.Type == "risk" || ev.Type == "summary" || ev.Type == "systems"
	})
}

// runRecommendationCall synthesizes a final recommendation from category summaries.
func runRecommendationCall(ctx context.Context, apiKey, model string, categoryEvents []StreamEvent) ([]StreamEvent, error) {
	var b strings.Builder
	b.WriteString("Based on the following category analysis, provide a final recommendation.\n\n")
	for _, ev := range categoryEvents {
		if id, ok := ev.Data["id"].(string); ok {
			summary, _ := ev.Data["summary"].(string)
			riskLevel, _ := ev.Data["riskLevel"].(string)
			fmt.Fprintf(&b, "## %s (risk: %s)\n%s\n\n", id, riskLevel, summary)
		}
	}

	sysPrompt := `You are JUST-PR. Based on the category analysis provided, emit exactly one JSON line:
{"type":"recommendation","data":{"action":"approve|request_changes|needs_review","reason":"<markdown: 2-3 sentences>"}}
Then emit: {"type":"done","data":{}}`

	return runStreamingCall(ctx, apiKey, model, sysPrompt, b.String(), func(ev StreamEvent) bool {
		return ev.Type == "recommendation"
	})
}

File: backend-ts/src/analysis/pipeline.ts
import { triage } from "./triage.js";
import { runSpecialist } from "./specialist.js";
import { analyzeWithAgentSDK } from "../providers/agent.js";
import { summarySystemPrompt } from "./prompt.js";
import type { PRFile, StreamEvent, ExistingComment, TriageResult, Emit } from "./types.js";
import { makeAnthropicClient, MODEL } from "../providers/anthropic-client.js";

function aboveThreshold(files: PRFile[], diff: string): boolean {
  if (files.length >= 2) return true;
  return diff.split("\n").length >= 50;
}

function first2000Lines(s: string): string {
  const lines = s.split("\n");
  if (lines.length <= 2000) return s;
  return lines.slice(0, 2000).join("\n") + "\n[truncated for summary]";
}

/** Run risk/summary/systems call (non-agent, no cwd needed). */
async function runSummaryCall(
  apiKey: string,
  triageResult: TriageResult,
  condensedDiff: string,
): Promise<StreamEvent[]> {
  void apiKey;
  const client = makeAnthropicClient();

  let prompt = "# Triage classification\n\n";
  for (const [cat, files] of Object.entries(triageResult)) {
    prompt += `- **${cat}**: ${files.join(", ")}\n`;
  }
  prompt += `\n# Diff (first 2000 lines)\n\n\`\`\`diff\n${condensedDiff}\n\`\`\`\n`;

  const stream = await client.messages.create({
    model: MODEL,
    max_tokens: 4096,
    system: summarySystemPrompt(),
    messages: [{ role: "user", content: prompt }],
    stream: true,
  });

  let buf = "";
  const events: StreamEvent[] = [];

  for await (const chunk of stream) {
    if (chunk.type === "content_block_delta" && chunk.delta.type === "text_delta") {
      buf += chunk.delta.text;
      const lines = buf.split("\n");
      buf = lines.pop() ?? "";
      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        try {
          const ev = JSON.parse(trimmed) as StreamEvent;
          if (["risk", "summary", "systems"].includes(ev.type)) events.push(ev);
        } catch { /* ignore */ }
      }
    }
  }
  return events;
}

async function runRecommendationCall(
  apiKey: string,
  categoryEvents: StreamEvent[],
): Promise<StreamEvent[]> {
  void apiKey;
  const client = makeAnthropicClient();

  let prompt = "Based on the following category analysis, provide a final recommendation.\n\n";
  for (const ev of categoryEvents) {
    const id = ev.data.id as string;
    const summary = ev.data.summary as string;
    const riskLevel = ev.data.riskLevel as string;
    prompt += `## ${id} (risk: ${riskLevel})\n${summary}\n\n`;
  }

  const sysPrompt = `You are PR-LENS. Based on the category analysis provided, emit exactly one JSON line:
{"type":"recommendation","data":{"action":"approve|request_changes|needs_review","reason":"<markdown: 2-3 sentences>"}}
Then emit: {"type":"done","data":{}}`;

  const msg = await client.messages.create({
    model: MODEL,
    max_tokens: 512,
    system: sysPrompt,
    messages: [{ role: "user", content: prompt }],
  });

  const events: StreamEvent[] = [];
  for (const block of msg.content) {
    if (block.type !== "text") continue;
    for (const line of block.text.split("\n")) {
      try {
        const ev = JSON.parse(line.trim()) as StreamEvent;
        if (ev.type === "recommendation") events.push(ev);
      } catch { /* ignore */ }
    }
  }
  return events;
}

const RISK_RANK: Record<string, number> = { critical: 0, high: 1, medium: 2, low: 3 };

export async function runPipeline(
  apiKey: string,
  diff: string,
  prFiles: PRFile[],
  cloneDir: string,
  existingComments: ExistingComment[],
  emit: Emit,
): Promise<void> {
  if (!aboveThreshold(prFiles, diff)) {
    // Small PR: run single agent session
    await analyzeWithAgentSDK(diff, cloneDir, existingComments, emit);
    return;
  }

  // Stage 1: triage
  let triageResult: TriageResult;
  try {
    triageResult = await triage(apiKey, prFiles);
    console.log(`[pipeline] triage: ${Object.keys(triageResult).length} categories`);
  } catch (err) {
    console.error(`[pipeline] triage failed, using fallback: ${err}`);
    await analyzeWithAgentSDK(diff, cloneDir, existingComments, emit);
    return;
  }

  if (Object.keys(triageResult).length === 0) {
    await analyzeWithAgentSDK(diff, cloneDir, existingComments, emit);
    return;
  }

  // Stage 2: kick off summary + all specialists in parallel. Each specialist
  // emits its category event as soon as it finishes, so the frontend sees
  // categories populate progressively instead of all appearing at the end.
  const summaryStart = Date.now();
  const specialistStart = Date.now();

  const summaryPromise = runSummaryCall(apiKey, triageResult, first2000Lines(diff))
    .then(async (events) => {
      console.log(`[pipeline] summary done in ${Date.now() - summaryStart}ms`);
      for (const ev of events) await emit(ev);
    })
    .catch((err) => { console.error("[pipeline] summary error:", err); });

  const seenCategories = new Set<string>();
  const allCategoryEvents: StreamEvent[] = [];

  const specialistPromises = Object.entries(triageResult).map(([cat, files]) => {
    const t = Date.now();
    return runSpecialist(cat, files, diff, cloneDir, existingComments)
      .then(async (events) => {
        console.log(`[pipeline] specialist ${cat} done in ${Date.now() - t}ms (${files.length} files)`);
        for (const ev of events) {
          if (ev.type !== "category") continue;
          const id = ev.data.id as string;
          if (id && seenCategories.has(id)) continue;
          if (id) seenCategories.add(id);
          allCategoryEvents.push(ev);
          await emit(ev);
        }
      })
      .catch((err) => { console.error(`[pipeline] specialist ${cat} error:`, err); });
  });

  // Wait for summary + all specialists to complete (events already emitted).
  await Promise.all([summaryPromise, ...specialistPromises]);
  console.log(`[pipeline] all specialists done in ${Date.now() - specialistStart}ms`);

  // Recommendation (after all categories are known).
  if (allCategoryEvents.length > 0) {
    // Sort by risk so the recommendation call sees highest-risk categories first.
    const ranked = [...allCategoryEvents].sort((a, b) => {
      const ra = RISK_RANK[a.data.riskLevel as string] ?? 3;
      const rb = RISK_RANK[b.data.riskLevel as string] ?? 3;
      return ra - rb;
    });
    const recEvents = await runRecommendationCall(apiKey, ranked)
      .catch((err) => { console.error("[pipeline] recommendation error:", err); return [] as StreamEvent[]; });
    for (const ev of recEvents) await emit(ev);
  }
}
